x-pack/plugins/ingest_manager/server/services/epm/archive/index.ts

```diff
@@ -0,0 +1,320 @@
+  const paths = await unpackArchiveToCache(archiveBuffer, contentType);
+  const archivePackageInfo = parseAndVerifyArchive(paths);
+  setArchiveFilelist(archivePackageInfo.name, archivePackageInfo.version, paths);
+
+  return {
+    paths,
+    archivePackageInfo,
+  };
+}
+
+function getBufferExtractorForContentType(contentType: string) {
+  if (contentType === 'application/gzip') {
+    return untarBuffer;
+  } else if (contentType === 'application/zip') {
+    return unzipBuffer;
+  } else {
+    throw new PackageUnsupportedMediaTypeError(
+      `Unsupported media type ${contentType}. Please use 'application/gzip' or 'application/zip'`
+    );
+  }
+}
+
+export async function unpackArchiveToCache(
+  archiveBuffer: Buffer,
+  contentType: string,
+  filter = (entry: ArchiveEntry): boolean => true
+): Promise<string[]> {
+  const bufferExtractor = getBufferExtractorForContentType(contentType);
+  const paths: string[] = [];
+  try {
+    await bufferExtractor(archiveBuffer, filter, (entry: ArchiveEntry) => {
+      const { path, buffer } = entry;
+      // skip directories
+      if (path.slice(-1) === '/') return;
+      if (buffer) {
+        cacheSet(path, buffer);
+        paths.push(path);
+      }
+    });
+  } catch (error) {
+    throw new PackageInvalidArchiveError(
+      `Error during extraction of uploaded package: ${error}. Assumed content type was ${contentType}, check if this matches the archive type.`
+    );
+  }
+
+  // While unpacking a tar.gz file with unzipBuffer() will result in a thrown error in the try-catch above,
+  // unpacking a zip file with untarBuffer() just results in nothing.
+  if (paths.length === 0) {
+    throw new PackageInvalidArchiveError(
+      `Uploaded archive seems empty. Assumed content type was ${contentType}, check if this matches the archive type.`
+    );
+  }
+  return paths;
+}
+
+// TODO: everything below performs verification of manifest.yml files, and hence duplicates functionality already implemented in the
+// package registry. At some point this should probably be replaced (or enhanced) with verification based on
+// https://github.com/elastic/package-spec/
+
+function parseAndVerifyArchive(paths: string[]): ArchivePackage {
+  // The top-level directory must match pkgName-pkgVersion, and no other top-level files or directories may be present
+  const toplevelDir = paths[0].split('/')[0];
+  paths.forEach((path) => {
+    if (path.split('/')[0] !== toplevelDir) {
+      throw new PackageInvalidArchiveError('Package contains more than one top-level directory.');
+    }
+  });
+
+  // The package must contain a manifest file ...
+  const manifestFile = `${toplevelDir}/manifest.yml`;
+  const manifestBuffer = cacheGet(manifestFile);
+  if (!paths.includes(manifestFile) || !manifestBuffer) {
+    throw new PackageInvalidArchiveError('Package must contain a top-level manifest.yml file.');
+  }
+
+  // ... which must be valid YAML
+  let manifest;
```

Comment:
Should this be packageManifest type?

Reply:
See my comment above -- this is the raw object parsed from yaml. Typing this would add a lot of code, but no further run-time security, as TypeScript does not exist at runtime and hence cannot guarantee run-time type safety.